Fix role assignment for <header> inside <main> and sectioning elements

joone opened the pull request:
#### da3bbf117740b8052a9e28fe42d9f6aadd283c56
<pre>
Fix role assignment for &lt;header&gt; inside &lt;main&gt; and sectioning elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=242685">https://bugs.webkit.org/show_bug.cgi?id=242685</a>
&lt;<a href="https://rdar.apple.com/48370244">rdar://48370244</a>&gt;

Reviewed by Tyler Wilcock.

This change assigns a generic role to &lt;header&gt; elements when nested
within &lt;main&gt;, &lt;aside&gt;, &lt;article&gt;, &lt;section&gt;, or &lt;nav&gt; elements, aligning with
the W3C spec(<a href="https://w3c.github.io/html-aam/#el-header)">https://w3c.github.io/html-aam/#el-header)</a>

Previously, the &lt;main&gt; element was not included in these conditions.
The &lt;header&gt; element assumed an unknown role instead of a generic role in these
conditions.

* LayoutTests/accessibility/mac/html-section-elements-expected.txt:
* LayoutTests/accessibility/mac/html-section-elements.html:
* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):

Canonical link: <a href="https://commits.webkit.org/273188@main">https://commits.webkit.org/273188@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64943fa286d12d5730a3e8ff87f4fd1d92bdb020

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34489 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37174 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31181 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10415 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30167 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11255 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30739 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9837 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9918 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30801 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38452 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31317 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31121 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35984 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10040 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33933 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11858 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30425 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7948 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10596 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10901 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/mac/html-section-elements.html:
```diff
@@ -44,8 +49,10 @@
 
           // When a header and footer are inside a section, they should not be landmarks.
-          // Instead, expect only the static text inside to be exposed.
-          shouldBe("body.childAtIndex(7).role", "'AXRole: AXStaticText'");
```
We probably still want to verify the role of `childAtIndex(7)` (which is the `section > header`). Seems like it becomes a group after this change, which is fine since `section > footer` is a group.

joone replied:
I think body.ChildAtIndex(7) is pointnig the ```<section>```. After this change, it becomes AXGroup.

twilco replied:
I ran this test with your change and added DOM `id`s, and I'm seeing `childAtIndex(7)` as `section > header`.

twilco commented on LayoutTests/accessibility/mac/html-section-elements.html:
```diff
@@ -44,8 +49,10 @@
           // When a header and footer are inside a section, they should not be landmarks.
-          // Instead, expect only the static text inside to be exposed.
-          shouldBe("body.childAtIndex(7).role", "'AXRole: AXStaticText'");
+          shouldBe("body.childAtIndex(7).childAtIndex(0).subrole", "'AXSubrole: '");
```
This is verifying the subrole of the static text within `section > header`, right? If so, that's probably not what we want to test (should be `childAtIndex(7).childAtIndex(0)`).

joone replied:
Do you want to check the value of childAtIndex(7).childAtIndex(0)? It returns a type of object.  I just wanted to check the subrole of section > header like checking article > header. Anyway, we need to check the role and subrole of both as you mentioned below.

twilco replied:
Sorry, my last comment was wrong (specifically the "should be" part). But the other part is accurate — `childAtIndex(7).childAtIndex(0)` is not the `section > header`, it's the static text AX object within `section > header`. If your intention is to verify the subrole of `section > header`, I believe `childAtIndex(7)` is the right one. Please add an `id` to this one and verify the `domIdentifer` to help future us understand this more easily.

joone replied:
Done.

twilco commented on LayoutTests/accessibility/mac/html-section-elements.html:
```diff
@@ -44,8 +49,10 @@
+          shouldBe("body.childAtIndex(7).childAtIndex(0).subrole", "'AXSubrole: '");
+
+          // When a header and footer are inside a main, they should not be landmarks.
+          shouldBe("body.childAtIndex(8).childAtIndex(0).subrole", "'AXSubrole: '");
```
This is verifying the `subrole` of static text within `section > footer`, right? If so, also probably not the thing we're trying to test.

I would expect the new testcases you added to be `childAtIndex(9)` and 10, right? If so, I don't think we are actually testing those in any way? Also it would be nice to verify their roles in addition to their subroles.

In general I don't love this test...it's hard to know what element `childAtIndex(x)` is supposed to be testing (of course this is not your fault, this problem existed before this patch). A nice improvement would be adding an `id` to the element, and then verifying the `domIdentifer()` in addition to the the role and subrole.

joone replied:
You are right. The numbering is confusing. How about fixing this including the role checking in a separate patch?: 
  * Adding unique ID to the elements
  * Test their roles and subroles.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -509,8 +509,11 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
-    if (node->hasTagName(headerTag) && !isDescendantOfElementType({ articleTag, asideTag, navTag, sectionTag }))
-        return AccessibilityRole::LandmarkBanner;
+    if (node->hasTagName(headerTag)) {
+        if (!isDescendantOfElementType({ articleTag, asideTag, mainTag, navTag, sectionTag }))
```
I'm confused why `header` only checks `articleTag, asideTag, mainTag, navTag, sectionTag`, while footer checks `articleTag, asideTag, navTag, sectionTag, mainTag, blockquoteTag, detailsTag, dialogTag, fieldsetTag, figureTag, tdTag`. Should these be the same?

joone replied:
Yes, they should check the same element types. I can fix this in a separate patch.

twilco reviewed (commented):
It would be really nice if there were a WPT for this, rather than only an internal layout test. That would help us be certain that we are doing the right thing.

twilco reviewed (commented):
As of this round of review, I don't think we are testing the new testcases (`main > header`, `main > footer`), which are `childAtIndex(9).childAtIndex(0)` and `childAtIndex(9).childAtIndex(1)`.

I'd prefer to fix this test as part of this patch, it's becoming too unwieldy to understand what is being tested (evidenced by this back and forth 🙂).

Please add `id`s to your new test cases (`main`, `main > header`, `main > footer`), and set up `shouldBe`s to verify the `domIdentifier`, `role`, and `subrole` of these (should be `childAtIndex(9)`, `childAtIndex(9).childAtIndex(0)`, `childAtIndex(9).childAtIndex(1)`).

Alternatively, I'd also be happy with a new WPT testcase, and using this patch to mark a new PASS. (We really need a new WPT testcase regardless)

joone commented on LayoutTests/accessibility/mac/html-section-elements.html:
```diff
@@ -41,11 +46,35 @@
+          shouldBe("body.childAtIndex(6).childAtIndex(1).role", "'AXRole: AXGroup'");
+          shouldBe("body.childAtIndex(6).childAtIndex(1).subrole", "'AXSubrole: AXFooter'");
+
+	  // FIXME: How can we access <section>?
```
The index value of childAtIndex(n) seems unusual. It counts all the children of the body element, but the `<section>` element doesn't have an index. How can we access the `<section>` element using the childAtIndex method?

twilco replied:
So this is actually using the `AccessibilityUIElement::childAtIndex` API (because `body` was retrieved via `accessibilityController.focusedElement`). This means it's using the accessibility tree children, not the DOM children, and the accessibility tree can differ quite a lot from the DOM.

What's very likely happening here is that the section element is determined to be "ignored" by accessibility (you can search for that term in the codebase to get a lot of context), which effectively means it's not part of the AX tree, and thus cannot be retrieved via `childAtIndex`. This is fine, I don't think we need to test section here (unless there's something I'm missing), so we can remove the FIXME.

twilco commented on LayoutTests/accessibility/mac/html-section-elements.html:
```diff
@@ -41,11 +46,35 @@
+
+          shouldBe("body.childAtIndex(8).domIdentifier", "'section_footer'");
+          shouldBe("body.childAtIndex(8).role", "'AXRole: AXGroup'");
+          // FIXME: AXSubrole of the footer should be empty.
```
Curious why you say this — is it because https://w3c.github.io/html-aam/#el-footer doesn't list a subrole (just "Use WAI-ARIA mapping")? Or some other reason?

joone replied:
Because the  footer element should have a generic role like the header element when used within the main element or a sectioning content element according to the spec.

twilco commented on LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:
```diff
@@ -40,6 +40,9 @@
   <header data-testname="el-header" aria-label="x" class="ex-generic">x</header>
 </nav>
 <header data-testname="el-header-ancestorbody" data-expectedrole="banner" class="ex">x</header>
+<main>
+  <header data-testname="el-header-ancestormain" data-expectedrole="generic" class="ex">x</header>
+</main>
```
You may be aware, but by adding a testcase to this WPT, we'll need to make sure to export it. The `Tools/Scripts/export-w3c-test-changes` should be able to help.

joone replied:
https://github.com/web-platform-tests/wpt/pull/44080

twilco reviewed (commented):
Thank you for improving this layout test!

twilco reviewed (approved):
Thank you @joone!

Commit:
Fix role assignment for <header> inside <main> and sectioning elements
https://bugs.webkit.org/show_bug.cgi?id=242685
<rdar://48370244>

Reviewed by Tyler Wilcock.

This change assigns a generic role to <header> elements when nested
within <main>, <aside>, <article>, <section>, or <nav> elements, aligning with
the W3C spec(https://w3c.github.io/html-aam/#el-header)

Previously, the <main> element was not included in these conditions.
The <header> element assumed an unknown role instead of a generic role in these
conditions.

* LayoutTests/accessibility/mac/html-section-elements-expected.txt:
* LayoutTests/accessibility/mac/html-section-elements.html:
* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.html:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):

Canonical link: https://commits.webkit.org/273188@main
